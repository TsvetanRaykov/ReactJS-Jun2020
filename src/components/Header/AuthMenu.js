import React, { useState, useContext, useEffect } from 'react'
import { withRouter } from 'react-router-dom'
import userService from '../../services/userService'
import {
	Typography,
	IconButton,
	Menu,
	MenuItem,
	Avatar,
	CircularProgress,
} from '@material-ui/core'

import UserContext from '../../Context'

const AuthMenu = (props) => {
	const [anchorEl, setAnchorEl] = useState(null)

	const { updateQuiz, user } = useContext(UserContext)
	const { userImg: img } = user || { userImg: '-' }
	const [userName, setUserName] = useState('')
	const [userImg, setUserImg] = useState('-')
	const [didMount, setDidMount] = useState(false)

	useEffect(() => {
		setUserImg(userService.getCurrentUser().userImg)
		setUserName(userService.getCurrentUser().userName)
		setDidMount(true)
	}, [])

	useEffect(() => {
		if (didMount) {
			setUserImg(img)
		}
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [img])

	const open = Boolean(anchorEl)

	const handleMenu = (event) => {
		setAnchorEl(event.currentTarget)
	}

	const handleClose = () => {
		setAnchorEl(null)
	}

	const handleAddQuiz = () => {
		updateQuiz(null)
		props.history.push('/quiz/edit')
	}

	const handleProfileClick = () => {
		props.history.push('/dashboard')
	}

	async function logout() {
		setAnchorEl(null)
		await userService.logout()
		props.history.push('/')
	}

	return (
		<div>
			<Typography component='span'>{userName}</Typography>
			<IconButton
				data-cy='authMenu'
				aria-label='account of current user'
				aria-controls='menu-appbar'
				aria-haspopup='true'
				onClick={handleMenu}
				color='inherit'
			>
				{userImg ? (
					<Avatar src={userImg} />
				) : (
					<CircularProgress color='secondary' />
				)}
			</IconButton>
			<Menu
				id='menu-appbar'
				anchorEl={anchorEl}
				anchorOrigin={{
					vertical: 'top',
					horizontal: 'right',
				}}
				keepMounted
				transformOrigin={{
					vertical: 'top',
					horizontal: 'right',
				}}
				open={open}
				onClose={handleClose}
			>
				<MenuItem onClick={handleProfileClick}>Profile</MenuItem>
				<MenuItem onClick={handleAddQuiz}>Add quiz</MenuItem>
				<MenuItem onClick={logout}>Logout</MenuItem>
			</Menu>
		</div>
	)
}

export default withRouter(AuthMenu)
