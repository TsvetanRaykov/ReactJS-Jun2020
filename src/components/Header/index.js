import React from 'react'
import withStyles from '@material-ui/core/styles/withStyles'
import {
	AppBar,
	Toolbar,
	Typography,
	Avatar,
	Box,
	Grid,
} from '@material-ui/core'
import AuthMenu from './AuthMenu'
import QuizTimer from '../../components/Timer'
import { withRouter } from 'react-router-dom'

const styles = (theme) => ({
	root: {
		flexGrow: 1,
	},
	logo: {
		marginRight: theme.spacing(2),
		cursor: 'pointer',
	},
	title: {
		flexGrow: 1,
	},
	timer: {
		fontSize: '30px',
		letterSpacing: '2px',
	},
})

const Header = (props) => {
	const { classes, title, timer, history } = props
	const handleLogoClick = () => {
		history.push('/dashboard')
	}
	return (
		<AppBar position='static' className={classes.root}>
			<Toolbar>
				<Grid container alignItems='center'>
					<Grid item xs={4}>
						<Box display='flex' alignItems='center'>
							<Avatar
								src='/logo.jpg'
								alt='quizoom'
								className={classes.logo}
								onClick={handleLogoClick}
							/>
							<Typography variant='h6' className={classes.title}>
								{`Quizoom${title ? ': ' + title : ''}`}
							</Typography>
						</Box>
					</Grid>
					<Grid item xs={4}>
						<Box
							display='flex'
							alignItems='center'
							justifyContent='center'
							className={classes.timer}
						>
							{timer && <QuizTimer timer={timer} />}
						</Box>
					</Grid>
					<Grid item xs={4}>
						<Box display='flex' alignItems='center' justifyContent='flex-end'>
							<AuthMenu />
						</Box>
					</Grid>
				</Grid>
			</Toolbar>
		</AppBar>
	)
}

export default withRouter(withStyles(styles)(Header))
