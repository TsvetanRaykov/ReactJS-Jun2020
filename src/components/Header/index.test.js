import React from 'react'
import { act } from 'react-test-renderer'
import { shallow } from 'enzyme'
import TestWrapper from '../../utils/test-wrapper'
import Header from '.'
import AuthMenu from './AuthMenu'

jest.mock('react', () => ({
	...jest.requireActual('react'),
	useLayoutEffect: jest.requireActual('react').useEffect,
}))

describe('<Header />', () => {
	it('Should match the snapshot', () => {
		act(() => {
			const wrapper = shallow(
				<TestWrapper>
					<Header />
				</TestWrapper>
			)
			expect(wrapper).toMatchSnapshot()
		})
	})
})

describe('<AuthMenu />', () => {
	it('Should match the snapshot', () => {
		act(() => {
			const wrapper = shallow(
				<TestWrapper>
					<AuthMenu />
				</TestWrapper>
			)
			expect(wrapper).toMatchSnapshot()
		})
	})
})
