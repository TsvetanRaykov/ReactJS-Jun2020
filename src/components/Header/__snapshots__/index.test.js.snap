// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthMenu /> Should match the snapshot 1`] = `ShallowWrapper {}`;

exports[`<Header /> Should match the snapshot 1`] = `ShallowWrapper {}`;
